File: src/main/java/com/amazonaws/intellij/ui/ui/widgets/AwsRegionPanel.java
package com.amazonaws.intellij.ui.ui.widgets;

import com.amazonaws.intellij.core.region.AwsRegion;
import com.amazonaws.intellij.core.region.AwsRegionManager;
import com.intellij.openapi.ui.ComboBox;
import com.intellij.ui.CollectionComboBoxModel;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionListener;
import java.util.Optional;

/**
 * Created by zhaoxiz on 7/28/17.
 */
public class AwsRegionPanel {
    private JPanel regionPanel;
    private ComboBox<AwsRegion> regionCombo;
    private final CollectionComboBoxModel<AwsRegion> regionModel = new CollectionComboBoxModel<>();

    public AwsRegionPanel(String defaultRegion) {
        regionCombo.setRenderer(new DefaultListCellRenderer() {
            @Override
            public Component getListCellRendererComponent(JList list, Object value, int index, boolean isSelected, boolean cellHasFocus) {
                JLabel label = (JLabel) super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus);
                label.setIcon(((AwsRegion) value).getIcon());
                return label;
            }
        });

        regionCombo.setModel(regionModel);

        AwsRegionManager.INSTANCE.getRegions().forEach(regionModel::add);
        selectRegion(defaultRegion);
    }

    public void addActionListener(ActionListener actionListener) {
        regionCombo.addActionListener(actionListener);
    }

    public JPanel getRegionPanel() {
        return regionPanel;
    }

    private void selectRegion(String regionId) {
        AwsRegionManager.INSTANCE.getRegions().stream().filter((region) -> region.getId().equals(regionId))
                .findFirst().ifPresent(regionModel::setSelectedItem);
    }

    public String getSelectedRegion() {
        assert regionModel.getSelected() != null;
        return regionModel.getSelected().getId();
    }
}

File: src/main/kotlin/com/amazonaws/intellij/ui/widgets/AwsRegionPanel.kt
package com.amazonaws.intellij.ui.widgets

import com.amazonaws.intellij.core.region.AwsRegion
import com.amazonaws.intellij.core.region.AwsRegionManager
import com.intellij.openapi.ui.ComboBox
import com.intellij.ui.CollectionComboBoxModel
import com.intellij.ui.ListCellRendererWrapper
import java.awt.FlowLayout
import java.awt.event.ActionListener
import javax.swing.JLabel
import javax.swing.JList
import javax.swing.JPanel

class AwsRegionPanel(private val defaultRegion: AwsRegion) {
    val regionPanel: JPanel = JPanel()
    private val regionCombo: ComboBox<AwsRegion> = ComboBox<AwsRegion>()
    private val regionModel = CollectionComboBoxModel<AwsRegion>(AwsRegionManager.regions.values.toList())

    init {
        regionCombo.renderer = object : ListCellRendererWrapper<AwsRegion>() {
            override fun customize(list: JList<*>, value: AwsRegion, index: Int, selected: Boolean, hasFocus: Boolean) {
                setIcon(value.icon)
            }
        }
        setupUI()
        regionCombo.model = regionModel
        regionModel.selectedItem = defaultRegion
    }

    fun addActionListener(actionListener: ActionListener) {
        regionCombo.addActionListener(actionListener)
    }

    fun getSelectedRegion(): AwsRegion? {
        return regionModel.selected
    }

    private fun setupUI() {
        regionPanel.layout = FlowLayout(FlowLayout.CENTER, 0, 0)
        val regionLabel = JLabel("Region")
        regionPanel.add(regionLabel)
        regionPanel.add(regionCombo)
        regionLabel.labelFor = regionCombo
    }
}